File: src/synapse_agent/tools/brave_search.py
"""Brave Search API — fallback search provider."""

from __future__ import annotations

import httpx

from synapse_agent.config import settings
from synapse_agent.state import SearchResult


async def brave_search(query: str, num_results: int = 5) -> list[SearchResult]:
    """Search via Brave Search API."""
    if not settings.brave_api_key:
        return []

    headers = {
        "Accept": "application/json",
        "Accept-Encoding": "gzip",
        "X-Subscription-Token": settings.brave_api_key,
    }
    params = {"q": query, "count": num_results}

    async with httpx.AsyncClient(timeout=30) as client:
        try:
            resp = await client.get(
                "https://api.search.brave.com/res/v1/web/search",
                headers=headers,
                params=params,
            )
            resp.raise_for_status()
            data = resp.json()
        except Exception:
            return []

    results = []
    for item in data.get("web", {}).get("results", [])[:num_results]:
        results.append(SearchResult(
            url=item.get("url", ""),
            title=item.get("title", ""),
            snippet=item.get("description", ""),
            source_query=query,
        ))
    return results

File: src/lib/synapse/tools/brave-search.ts
/**
 * Brave Search API — fallback search provider.
 * Port of Python's tools/brave_search.py
 */

import type { SearchResult } from "../types";
import { getSettings } from "../config";

export async function braveSearch(
  query: string,
  numResults = 5
): Promise<SearchResult[]> {
  const { braveApiKey } = getSettings();
  if (!braveApiKey) return [];

  const params = new URLSearchParams({
    q: query,
    count: String(numResults),
  });

  try {
    const resp = await fetch(
      `https://api.search.brave.com/res/v1/web/search?${params}`,
      {
        headers: {
          Accept: "application/json",
          "Accept-Encoding": "gzip",
          "X-Subscription-Token": braveApiKey,
        },
        signal: AbortSignal.timeout(30_000),
      }
    );

    if (!resp.ok) return [];
    const data = await resp.json();

    return (data.web?.results ?? [])
      .slice(0, numResults)
      .map(
        (item: Record<string, string>): SearchResult => ({
          url: item.url ?? "",
          title: item.title ?? "",
          snippet: item.description ?? "",
          content: "",
          relevanceScore: 0,
          sourceQuery: query,
        })
      );
  } catch {
    return [];
  }
}
